From 105728cf065d709d1db479778f3a51f3ef763679 Mon Sep 17 00:00:00 2001
From: Gregory Norton <gregory@serenade.local>
Date: Sat, 30 Aug 2026 00:00:00 +0000
Subject: [PATCH 2/7] Lagom: Build the full LibGUI library on host systems

The Lagom build only compiled six LibGUI files, which is not enough to
link GUI applications or WindowServer. Add GUI (and its Config dependency)
to the standard library list, drop the partial stub target, and generate
the NotificationServer IPC endpoint headers that LibGUI consumes but whose
generating service the host build does not include.

diff --git a/Userland/Libraries/LibGUI/CMakeLists.txt b/Userland/Libraries/LibGUI/CMakeLists.txt
index d20bbc80d4..7dd44476b7 100644
--- a/Userland/Libraries/LibGUI/CMakeLists.txt
+++ b/Userland/Libraries/LibGUI/CMakeLists.txt
@@ -1,5 +1,14 @@
 compile_gml(DynamicWidgetContainerControls.gml DynamicWidgetContainerControls.cpp)
 
+if (NOT SERENITYOS)
+    # The Serenity build generates these in the respective service directories;
+    # the Lagom host build does not build those services, so generate them here.
+    # (WindowServer's and Clipboard's endpoints are generated by their own
+    # service directories, which the Lagom build does include.)
+    compile_ipc(../../Services/NotificationServer/NotificationClient.ipc ../../Services/NotificationServer/NotificationClientEndpoint.h)
+    compile_ipc(../../Services/NotificationServer/NotificationServer.ipc ../../Services/NotificationServer/NotificationServerEndpoint.h)
+endif()
+
 compile_gml(AboutDialog.gml AboutDialogGML.cpp)
 stringify_gml(DatePickerDialog.gml DatePickerDialogGML.h date_picker_dialog_gml)
 compile_gml(EmojiInputDialog.gml EmojiInputDialogGML.cpp)
diff --git a/Meta/Lagom/CMakeLists.txt b/Meta/Lagom/CMakeLists.txt
index 213a1e77dd..94665e42be 100644
--- a/Meta/Lagom/CMakeLists.txt
+++ b/Meta/Lagom/CMakeLists.txt
@@ -468,6 +468,7 @@ if (BUILD_LAGOM)
         Archive
         Audio
         Compress
+        Config
         Crypto
         Cpp
         DeviceTree
@@ -478,6 +479,7 @@ if (BUILD_LAGOM)
         GL
         GLSL
         GPU
+        GUI
         HID
         HTTP
         IMAP
@@ -544,20 +547,6 @@ if (BUILD_LAGOM)
         add_serenity_subdirectory("Userland/Libraries/Lib${lib}")
     endforeach()
 
-    # GUI
-    set(LIBGUI_SOURCES
-        GML/Lexer.cpp
-        GML/Parser.cpp
-        GML/SyntaxHighlighter.cpp
-        Icon.cpp
-        Model.cpp
-        ModelIndex.cpp
-    )
-    list(TRANSFORM LIBGUI_SOURCES PREPEND "${SERENITY_PROJECT_ROOT}/Userland/Libraries/LibGUI/")
-    lagom_lib(LibGUI gui
-        SOURCES ${LIBGUI_SOURCES}
-        LIBS LibCore LibGfx LibSyntax)
-
     # FIXME: How about we don't include Kernel/API from random high-level libraries?
     install(FILES ${SERENITY_PROJECT_ROOT}/Kernel/API/KeyCode.h DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/Kernel/API")
 
